src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+namespace Roslyn.Utilities
+{
+    internal struct SharedStopwatch
+    {
+        private static readonly Stopwatch s_stopwatch = Stopwatch.StartNew();
+
+        private TimeSpan _accumulated;
+        private TimeSpan _started;
+
+        public readonly TimeSpan Elapsed
+        {
+            get
+            {
+                if (IsRunning)
+                    return _accumulated + s_stopwatch.Elapsed - _started;
+
+                return _accumulated;
+            }
+        }
+
+        public readonly long ElapsedMilliseconds => (long)Elapsed.TotalMilliseconds;
+        public readonly bool IsRunning => _started != TimeSpan.Zero;
+
+        public static SharedStopwatch StartNew()
+        {
+            var result = new SharedStopwatch();
+            result.Start();
+            return result;
+        }
+
+        public void Reset()
+        {
+            Stop();
+            _accumulated = TimeSpan.Zero;
+        }
+
+        public void Restart()
+        {
+            Reset();
+            Start();
+        }
+
+        public void Start()
+        {
+            if (!IsRunning)
+            {
+                _started = s_stopwatch.Elapsed;
+            }
+        }
+
+        public void Stop()
+        {
+            if (IsRunning)
+            {
+                _accumulated += s_stopwatch.Elapsed - _started;
+                _started = TimeSpan.Zero;
+            }
+        }
```

Comment:
Are these methods needed? It looks like the only method used is `Start()` above which could be replaced by `_started = ...;`.

Reply:
I was stuck deciding between drop-in full functionality and limited functionality for scenarios most impacted by allocation performance. There didn't seem to be much downside to the former considering the simplicity of the API.

Reply:
We should keep it simple. Could the type could be reduced to two `static` methods?
```C#
static long StartNew() => s_stopwatch.ElapsedTicks;

static long Stop(long start) => s_stopwatch.ElapsedTicks - start;
```

---
In reply to: [371412091](https://github.com/dotnet/roslyn/pull/41205#discussion_r371412091) [](ancestors = 371412091)

Reply:
Stopwatch itself exposes static APIs that return the raw tick count and frequency
in which case you don't even need a stopwatch instance, you just need to normalize to 100ns ticks

Reply:
Reduced the API to just `StartNew()` and `Elapsed`